application/controllers/collection_instrument.py

```
@@ -277,17 +278,43 @@ def unlink_instrument_from_exercise(self, instrument_id, exercise_id, session=No
-        for business in instrument.businesses:
-            bound_logger.info("Removing business/exercise link", business_id=business.id, ru_ref=business.ru_ref)
-            business.instruments.remove(instrument)
+        if instrument.type == "SEFT":
+            # SEFT instruments need to be deleted from both GCP and the db, removing just the link will leave orphaned
+            # data. When deleting SEFT instruments, the link will automatically be removed without this function
+            raise RasError(
+                f"{instrument_id} is of type SEFT which should be deleted and not unlinked",
+                405,
+            )
 
+        instrument.exercises.remove(exercise)
         response = self.publish_remove_collection_instrument(exercise_id, instrument.instrument_id)
         if response.status_code != 200:
             raise RasError("Failed to publish upload message", 500)
         bound_logger.info("Successfully unlinked instrument to exercise")
         return True
 
+    @with_db_session
+    def delete_seft_collection_instrument(self, instrument_id: str, session: Session = None) -> None:
+        """
+        Deletes a seft collection instrument from the database and gcs
+        :param instrument_id: A collection instrument id (UUID)
+        :param session: database session
+        """
+        instrument = self.get_instrument_by_id(instrument_id, session)
+
+        if not instrument:
+            raise RasError(f"Collection instrument {instrument_id} not found", 404)
+        if instrument.type != "SEFT":
+            raise RasError(
+                f"Only SEFT collection instruments can be deleted {instrument_id} has type {instrument.type}",
+                405,
+            )
+
+        session.delete(instrument)
+        gcs_seft_bucket = GoogleCloudSEFTCIBucket(current_app.config)
+        file_path = self._build_seft_file_path(instrument)
+        gcs_seft_bucket.delete_file_from_bucket(file_path)
```

It's debatable whether this should raise an exception if it fails, I have left as is, but if it can't find the blob is just writing an error in the logs enough?

---

Is the file not existing an issue ? it should be very rare the file not being there to delete. I think the unlinking / deleting should be enough since there would be the asumption that if they were correcting a mistake a few file would be loaded.

---

I ummed and arg on that one as well, the delete_file_from_bucket writes out to the logs which is at least something I guess. I came to the conclusion that I didn't know the current state of the data so it is completely feasible to have a record in the table and not in GCS. I also thought as long as the instrument is removed from the db and thus the UI we can handle orphaned data in GCS if we have to.